### ISSUE TITLE
how to record loss with tf.summary in the tf.function graph mode

### ISSUE THREAD

**Issue by sjtusmartboy** [tensorflow/tensorflow#51584]
**tensorflow 2.5, python 3.7**

```
import tensorflow as tf
import numpy as np
import datetime

class Dense(tf.Module):
    def __init__(self, input_dim, output_size, name=None):
        super(Dense, self).__init__(name=name)
        self.w = tf.Variable(
            tf.random.normal([input_dim, output_size]), name='w')
        self.b = tf.Variable(tf.zeros([output_size]), name='b')

    def __call__(self, x):
        y = tf.matmul(x, self.w) + self.b
        return tf.nn.relu(y)

model = Dense(2,4)

class Test(object):
    def __init__(self):
        self.output = model([[7.0, 3]])
        self.optimizer = tf.compat.v1.train.AdamOptimizer(0.4)

        self.step = 0

        stamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
        logdir = 'logs/test8/%s' % stamp

        self.summary_writer = tf.summary.create_file_writer(logdir)
        tf.summary.trace_on(graph=True, profiler=False)

        for i in range(10):
            self.run()

    def compare(self, y_true, output):
        return tf.square(y_true - output)

    def loss_fn(self):
        y_true = tf.ones([1,4])
        self.output = model([[7.0, 3]])

        comp = tf.py_function(self.compare, [y_true , self.output], tf.float32)
        loss = tf.reduce_mean(comp) # error

        # output2 = model([[7.0, 3]])
        # loss = tf.reduce_mean(tf.square(y_true - output2))

        tf.print(loss)
        tf.print(self.step)

        with self.summary_writer.as_default():
            tf.summary.scalar('loss', loss, step=self.step)

        self.step += 1
        return loss

    # @tf.function(autograph=True, jit_compile=True)
    @tf.function()
    def run(self):
        train_op = self.optimizer.minimize(self.loss_fn)

Test()
```

As can be seen from the output, the variable `self.step` is not updated in `def loss_fn(self)`. The reason is simple from what is said in https://tensorflow.google.cn/tensorboard/migrate, 

```
The "step" value must be passed into each op via a the step argument

TensorBoard requires a step value to render the data as a time series
Explicit passing is necessary because the global step from TF 1.x has been removed, so each op must know the desired step variable to read
```

However, according to https://tensorflow.google.cn/api_docs/python/tf/compat/v1/train/AdamOptimizer?hl=en&version=nightly#expandable-3, 

`When eager execution is enabled, loss should be a Python function that takes no arguments and computes the value to be minimized`.

Definitely these two announcement contradict to each other when trying to use `tf.summary` in `def loss_fn(self)`. Any solutions? I just want to record the loss with `tf.summary`.

**Comment by sjtusmartboy** [comment]
Problem solved by changing the code to graph mode
